internal/util/diff/diff_test.go

```diff
@@ -242,6 +254,115 @@ func TestCommand_Diff(t *testing.T) {
 	}
 }
 
+func TestCommand_InvalidRef(t *testing.T) {
+	reposChanges := map[string][]testutil.Content{
+		testutil.Upstream: {
+			{
+				Data:   testutil.Dataset2,
+				Branch: "master",
+				Tag:    "v2",
+			},
+			{
+				Data: testutil.Dataset3,
+			},
+		},
+	}
+
+	g := &testutil.TestSetupManager{
+		T:            t,
+		ReposChanges: reposChanges,
+		GetRef:       "v2",
+	}
+	defer g.Clean()
+
+	if !g.Init() {
+		return
+	}
+
+	diffOutput := &bytes.Buffer{}
+	err := (&Command{
+		Path:         g.LocalWorkspace.FullPackagePath(),
+		Ref:          "hurdygurdy", // ref should not exist in upstream
+		DiffType:     DiffTypeCombined,
+		DiffTool:     "diff",
+		DiffToolOpts: "-r -i -w",
+		Output:       diffOutput,
+	}).Run(fake.CtxWithNilPrinter())
+	assert.Error(t, err)
+}
+
+// Validate that all three directories are staged and provided to diff command
+func TestCommand_Diff3Parameters(t *testing.T) {
+	reposChanges := map[string][]testutil.Content{
+		testutil.Upstream: {
+			{
+				Data:   testutil.Dataset2,
+				Branch: "master",
+				Tag:    "v2",
+			},
+			{
+				Data: testutil.Dataset3,
+			},
+		},
+	}
+
+	g := &testutil.TestSetupManager{
+		T:            t,
+		ReposChanges: reposChanges,
+		GetRef:       "v2",
+	}
+	defer g.Clean()
+
+	if !g.Init() {
+		return
+	}
+
+	diffOutput := &bytes.Buffer{}
+	err := (&Command{
+		Path:         g.LocalWorkspace.FullPackagePath(),
+		Ref:          "master",
+		DiffType:     DiffType3Way,
+		DiffTool:     "echo", // this is a proxy for 3 way diffing to validate we pass proper values
+		DiffToolOpts: "",
+		Output:       diffOutput,
+	}).Run(fake.CtxWithNilPrinter())
+	assert.NoError(t, err)
+
+	// Expect 3 value to be printed (1 per source)
+	assert.Equal(t, 3, len(strings.Split(diffOutput.String(), " ")))
+}
+
+// Tests against directories in different states
+func TestCommand_NotAKptDirectory(t *testing.T) {
+	// Initial test setup
+	dir, err := ioutil.TempDir("", "example")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	testCases := map[string]struct {
+		directory string
+	}{
+		"Directory Is Not Kpt Package": {directory: dir},
+		"Directory Does Not Exist":     {directory: "/not/a/directory"},
+	}
+
+	for tn, tc := range testCases {
+		t.Run(tn, func(t *testing.T) {
+			diffOutput := &bytes.Buffer{}
+			cmdErr := (&Command{
+				Path:         tc.directory,
+				Ref:          "master",
+				DiffType:     DiffTypeCombined,
+				DiffTool:     "diff",
+				DiffToolOpts: "-r -i -w",
+				Output:       diffOutput,
+			}).Run(fake.CtxWithNilPrinter())
+			assert.Error(t, cmdErr)
```

Comment:
Same as above. When we expect errors, we should try to verify that it fails for the expected reason.